File: android/app/src/main/java/dev/pepotech/pepomote/ui/screens/ControllerLandscapeScreen.kt
package dev.pepotech.pepomote.ui.screens

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.displayCutoutPadding
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.BoxWithConstraints
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.statusBarsPadding
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.DisposableEffect
import dev.pepotech.pepomote.service.Route
import dev.pepotech.pepomote.sensor.Frame
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalView
import androidx.compose.ui.unit.dp
import dev.pepotech.pepomote.control.ButtonState
import dev.pepotech.pepomote.service.LinkState
import dev.pepotech.pepomote.service.UiLink
import dev.pepotech.pepomote.ui.components.CollapsibleHeader
import dev.pepotech.pepomote.ui.components.KeyboardDialog
import dev.pepotech.pepomote.ui.components.LocalPressRegistry
import dev.pepotech.pepomote.ui.components.NoticeBanner
import dev.pepotech.pepomote.ui.components.PadCross
import dev.pepotech.pepomote.ui.components.NearPill
import dev.pepotech.pepomote.ui.components.PrecisionPill
import dev.pepotech.pepomote.ui.components.RotateSuggestion
import dev.pepotech.pepomote.ui.components.RoundButton
import dev.pepotech.pepomote.ui.components.UiScale
import dev.pepotech.pepomote.ui.components.rememberPressRegistry
import dev.pepotech.pepomote.ui.components.slideCanvas
import dev.pepotech.pepomote.ui.theme.PepoColors
import androidx.compose.ui.res.stringResource
import dev.pepotech.pepomote.R
import kotlin.math.roundToInt

/**
 * Mando apaisado estilo "de lado" (NES): cruceta a la izquierda, 1 y 2
 * grandes a la derecha. Para juegos 2D en Dolphin con el Wiimote de lado.
 * `showChips`: selector Puntero/Dolphin/Wii U, mismas condiciones que en
 * vertical. Dentro de Wii U como Mando de Wii: cabecera «Wii U · Mando de
 * Wii» con «Teclado» (texto para el teclado en pantalla de Cemu) y el
 * selector «En Cemu soy» debajo.
 */
@Composable
fun ControllerLandscapeScreen(link: UiLink, showChips: Boolean, onDisconnect: () -> Unit) {
    val view = LocalView.current
    val context = androidx.compose.ui.platform.LocalContext.current
    var keyboardOpen by remember { mutableStateOf(false) }
    DisposableEffect(Unit) {
        view.keepScreenOn = true
        onDispose { view.keepScreenOn = false }
    }
    // Como en vertical: el teclado y la pose congelada van juntos y se sueltan
    // solos (al cerrar, al cambiar de modo, al reconectar o al salir)
    androidx.compose.runtime.LaunchedEffect(keyboardOpen, link) {
        if (keyboardOpen && !dev.pepotech.pepomote.service.Route.showsKeyboard(link)) keyboardOpen = false
        LinkState.motion?.pointerHold =
            keyboardOpen && dev.pepotech.pepomote.service.Route.holdsPointerForKeyboard(link)
    }
    DisposableEffect(Unit) { onDispose { LinkState.motion?.pointerHold = false } }
    // Móvil de lado: los sensores giran con el mando (Route.sidewaysRotation:
    // mando girado con el IR a la izquierda en un juego, o apuntando con el
    // borde largo en modo puntero); al salir, el mando vertical de siempre
    val engine = LinkState.motion
    val rotation = rememberDisplayRotation()
    LaunchedEffect(engine, link, rotation) {
        engine?.rotation = Route.sidewaysRotation(link, rotation)
    }
    DisposableEffect(engine) {
        onDispose {
            engine?.rotation = Frame.ROTATION_0
            ButtonState.reset()
        }
    }

    // Capa de pulsación de la pantalla: cada botón dice dónde está y, con
    // «Pulsar deslizando», el dedo pasa de uno a otro sin levantarlo
    val press = rememberPressRegistry()
    DisposableEffect(press) { onDispose { press.releaseAll() } }

    CompositionLocalProvider(LocalPressRegistry provides press) {
        BoxWithConstraints(
            modifier = Modifier
                .fillMaxSize()
                .background(PepoColors.Background)
                .statusBarsPadding()
                .displayCutoutPadding()
                .navigationBarsPadding()
                .slideCanvas()
        ) {
            // En una tablet todo crece a la vez (UiScale; en cualquier móvil, 1)
            val s = UiScale.landscape(maxWidth.value, maxHeight.value)

            // Cruceta izquierda
            Box(
                modifier = Modifier
                    .align(Alignment.CenterStart)
                    .padding(start = 34.dp * s)
            ) {
                // En un juego, la cruceta de un mando girado (IR a la izquierda)
                PadCross(sizeDp = 190.dp * s, sideways = Route.sidewaysDpad(link))
            }

            // RetroArch como mando de NES: 1 y 2 son B y A del RetroPad, la A
            // grande es X y Home abre el menú de RetroArch (el receptor los mapea)
            val retro = Route.isRetroArch(link)
            // − / ◎ / + y A en el centro (un 20 % más grandes que en la primera
            // versión: de lado se pulsan con el pulgar y quedaban pequeños). La
            // diana, como en el mando vertical y en el mando + Nunchuk: mantener
            // recentra el cursor (puntero) o el puntero IR (Dolphin, Mando de Wii
            // en Cemu), que de lado también se apunta. Home, pequeño, a la derecha
            // de A y fuera del modo puntero (el PC no le da uso): como en el Mando
            // de Wii de lado, donde −/Home/+ quedan entre A y 1/2; Dolphin abre con
            // él el menú HOME y Mario Party 10 lo pide para emparejar cada mando
            Column(
                modifier = Modifier.align(Alignment.Center),
                horizontalAlignment = Alignment.CenterHorizontally,
                verticalArrangement = Arrangement.spacedBy(12.dp * s)
            ) {
                Spacer(Modifier.height(20.dp * s))
                Row(
                    horizontalArrangement = Arrangement.spacedBy(16.dp * s),
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    RoundButton("−", 53.dp * s, ButtonState.MINUS, textSize = (19 * s).roundToInt())
                    RecenterButton(size = 54.dp * s)
                    RoundButton("+", 53.dp * s, ButtonState.PLUS, textSize = (19 * s).roundToInt())
                }
                if (showHomeButton(link)) {
                    // un hueco igual a la izquierda deja A centrada bajo la diana
                    Row(
                        horizontalArrangement = Arrangement.spacedBy(14.dp * s),
                        verticalAlignment = Alignment.CenterVertically
                    ) {
                        Spacer(Modifier.size(40.dp * s))
                        RoundButton(if (Route.retroNes(link)) "X" else "A", 62.dp * s, ButtonState.A, textSize = (22 * s).roundToInt())
                        RoundButton(stringResource(if (retro) R.string.retro_menu else R.string.home_btn), 40.dp * s, ButtonState.HOME, textSize = ((if (retro) 10 else 11) * s).roundToInt())
                    }
                } else {
                    RoundButton("A", 62.dp * s, ButtonState.A, textSize = (22 * s).roundToInt())
                }
            }

            // 1 y 2 grandes a la derecha (los botones de acción del modo NES)
            Row(
                modifier = Modifier
                    .align(Alignment.CenterEnd)
                    .padding(end = 30.dp * s),
                horizontalArrangement = Arrangement.spacedBy(18.dp * s)
            ) {
                RoundButton(
                    if (retro) "B" else "1", 92.dp * s, ButtonState.ONE,
                    background = PepoColors.Blue,
                    pressedColor = PepoColors.BlueHover,
                    textColor = PepoColors.OnAccent,
                    textSize = (28 * s).roundToInt()
                )
                RoundButton(
                    if (retro) "A" else "2", 92.dp * s, ButtonState.TWO,
                    background = PepoColors.Blue,
                    pressedColor = PepoColors.BlueHover,
                    textColor = PepoColors.OnAccent,
                    textSize = (28 * s).roundToInt(),
                    pop = true
                )
            }

            if (retro) {
                Box(Modifier.align(Alignment.BottomCenter).padding(bottom = 8.dp).width(112.dp * s)) {
                    dev.pepotech.pepomote.ui.components.TriggerZone(
                        label = if (Route.retroNes(link)) "Y" else "B", height = 44.dp * s)
                }
            } else Text(
                stringResource(R.string.rotate_hint),
                style = MaterialTheme.typography.bodyMedium,
                modifier = Modifier
                    .align(Alignment.BottomCenter)
                    .padding(bottom = 8.dp)
            )

            // En Dolphin la precisión no hace nada: la píldora es «Acercar»
            val pill = Modifier
                .align(Alignment.BottomStart)
                .padding(start = 16.dp, bottom = 6.dp)
            if (isDolphin(link)) NearPill(modifier = pill) else PrecisionPill(modifier = pill)

            // Aviso de «sin giroscopio real» (una vez, en modo puntero: de lado
            // también se apunta) y, debajo, los avisos transitorios del receptor
            Column(
                Modifier
                    .align(Alignment.TopCenter)
                    .padding(top = 84.dp),
                horizontalAlignment = Alignment.CenterHorizontally,
                verticalArrangement = Arrangement.spacedBy(8.dp)
            ) {
                RotateSuggestion()
                GyroWarnCard(link)
                NoticeBanner()
            }

            if (keyboardOpen) {
                KeyboardDialog(
                    onSend = { LinkState.sendText?.invoke(it) },
                    onClose = { keyboardOpen = false },
                    retroPad = Route.isRetroArch(link),
                    pointer = (link as? UiLink.Connected)?.mode == LinkState.MODE_POINTER
                )
            }

            // Cabecera plegable, la última del Box: desplegada tapa a los
            // botones (y no al revés) y a los cuatro segundos se quita de en
            // medio. Dentro lleva lo que antes iba en una línea: el PC, los
            // chips, «Teclado», «En Cemu soy» y «Salir»
            CollapsibleHeader(
                link = link,
                showChips = showChips,
                modifier = Modifier.align(Alignment.TopCenter),
                onKeyboard = {
                    dev.pepotech.pepomote.ui.components.openKeyboard(context, link, press)
                    keyboardOpen = true
                },
                onDisconnect = onDisconnect
            )
        }
    }
}

File: ios/PepoMote/UI/Screens/ControllerLandscapeScreen.swift
import SwiftUI

/// Medidas del mando apaisado (NES): en un iPad todo crece a la vez
/// (UiScale, por la anchura natural del trazado); en cualquier iPhone, las
/// de siempre.
struct LandscapeMetrics {
    let s: CGFloat
    let cross: CGFloat
    let small: CGFloat
    /// Diana de recentrado entre − y + (como en el mando + Nunchuk).
    let recenter: CGFloat
    let a: CGFloat
    /// Home, pequeño, a la derecha de A (fuera del modo puntero).
    let home: CGFloat
    let big: CGFloat
    let crossInset: CGFloat
    let bigInset: CGFloat

    init(size: CGSize) {
        s = UiScale.factor(size, base: UiScale.landscapeBase)
        cross = 190 * s
        small = 53 * s
        recenter = 54 * s
        a = 62 * s
        home = 40 * s
        big = 92 * s
        crossInset = 34 * s
        bigInset = 30 * s
    }

    func text(_ base: CGFloat) -> CGFloat { base * s }
    func gap(_ base: CGFloat) -> CGFloat { base * s }
}

/// Mando apaisado estilo "de lado" (NES): cruceta a la izquierda, 1 y 2
/// grandes a la derecha. Para juegos 2D en Dolphin con el Wiimote de lado.
/// Arriba, la cabecera plegable: una pastilla con el modo y, desplegada, la
/// tarjeta con el PC, «Teclado», los chips, el selector «En Cemu soy» (dentro
/// de Wii U como Mando de Wii) y «Salir».
struct ControllerLandscapeScreen: View {
    let showChips: Bool
    let onDisconnect: () -> Void
    @ObservedObject var link = LinkState.shared
    @State private var keyboardOpen = false
    /// Home del Mando de Wii: conectado y fuera del modo puntero.
    private var showHome: Bool { link.link.connected.map(showHomeButton) ?? false }
    /// RetroArch como mando de NES: 1 y 2 son B y A del RetroPad, la A grande
    /// es X y Home abre el menú de RetroArch (el receptor los mapea).
    private var retro: Bool { Route.isRetroArch(link.link) }
    @State private var rotation: Int = OrientationLock.frameRotation(OrientationLock.current)
    /// Dónde está cada botón, para «Pulsar deslizando».
    @StateObject private var press = PressRegistry()

    var body: some View {
        GeometryReader { geo in
            let m = LandscapeMetrics(size: geo.size)
            ZStack {
                // El primero: recoge los dedos que nacen fuera de los botones
                SlideCanvas()

                // Cruceta izquierda
                HStack {
                    // En un juego, la cruceta de un mando girado (IR a la izquierda)
                    PadCross(size: m.cross, sideways: Route.sidewaysDpad(link.link)).padding(.leading, m.crossInset)
                    Spacer()
                }

                // − / ◎ / + y A en el centro: la diana (mantener) recentra
                // el cursor o el puntero IR, que de lado también se apunta.
                // Home, pequeño, a la derecha de A y fuera del modo puntero
                // (el PC no le da uso): como en el Mando de Wii de lado, donde
                // −/Home/+ quedan entre A y 1/2; un hueco igual a la izquierda
                // deja A centrada bajo la diana
                VStack(spacing: m.gap(12)) {
                    Spacer().frame(height: m.gap(20))
                    HStack(spacing: m.gap(16)) {
                        RoundButton(label: "−", size: m.small, bit: Btn.minus, textSize: m.text(19))
                        RecenterButton(size: m.recenter)
                        RoundButton(label: "+", size: m.small, bit: Btn.plus, textSize: m.text(19))
                    }
                    if showHome {
                        HStack(spacing: m.gap(14)) {
                            Color.clear.frame(width: m.home, height: m.home)
                            RoundButton(label: retro ? "X" : "A", size: m.a, bit: Btn.a, textSize: m.text(22))
                            RoundButton(label: tr(retro ? "retro_menu" : "home_btn"), size: m.home, bit: Btn.home, textSize: m.text(retro ? 10 : 11))
                        }
                    } else {
                        RoundButton(label: retro ? "X" : "A", size: m.a, bit: Btn.a, textSize: m.text(22))
                    }
                }

                // 1 y 2 grandes a la derecha (los botones de acción del modo NES)
                HStack(spacing: m.gap(18)) {
                    Spacer()
                    RoundButton(label: retro ? "B" : "1", size: m.big, bit: Btn.one, background: Pepo.blue, pressedColor: Pepo.blueHover, textColor: Pepo.onAccent, textSize: m.text(28))
                    RoundButton(label: retro ? "A" : "2", size: m.big, bit: Btn.two, background: Pepo.blue, pressedColor: Pepo.blueHover, textColor: Pepo.onAccent, textSize: m.text(28), pop: true)
                }
                .padding(.trailing, m.bigInset)

                VStack {
                    Spacer()
                    Text(tr("rotate_hint")).pepoBody().lineLimit(1).padding(.bottom, 8)
                }

                VStack {
                    Spacer()
                    HStack {
                        // En Dolphin la precisión no hace nada: la píldora es «Acercar»
                        if link.link.connected?.mode == LinkState.modeDolphin {
                            NearPill().padding(.leading, 16).padding(.bottom, 6)
                        } else {
                            PrecisionPill().padding(.leading, 16).padding(.bottom, 6)
                        }
                        Spacer()
                    }
                }

                VStack {
                    NoticeBanner().padding(.top, 84)
                    Spacer()
                }
            }
            .frame(width: geo.size.width, height: geo.size.height)
            // La última: la tarjeta desplegada gana a la cruceta y a 1/2
            .overlay(alignment: .top) { header }
        }
        .coordinateSpace(name: PressRegistry.padSpace)
        .environmentObject(press)
        .background(Pepo.background.ignoresSafeArea())
        .sheet(isPresented: $keyboardOpen) {
            KeyboardSheet { keyboardOpen = false }
        }
        // Móvil de lado: los sensores giran con el mando (Route.sidewaysRotation:
        // mando girado con el IR a la izquierda en un juego, o apuntando con el
        // borde largo en modo puntero); al salir, el mando vertical de siempre
        .onChange(of: rotation) { _ in applyEngine() }
        .onChange(of: link.link) { _ in applyEngine() }
        .onReceive(NotificationCenter.default.publisher(for: UIDevice.orientationDidChangeNotification)) { _ in
            rotation = OrientationLock.frameRotation(OrientationLock.current)
        }
        .onAppear {
            UIApplication.shared.isIdleTimerDisabled = true
            UIDevice.current.beginGeneratingDeviceOrientationNotifications()
            rotation = OrientationLock.frameRotation(OrientationLock.current)
            applyEngine()
        }
        .onDisappear {
            UIApplication.shared.isIdleTimerDisabled = false
            link.motion?.rotation = Frame.rotation0
            press.releaseAll() // nada queda pulsado al salir
            ButtonState.shared.reset()
        }
    }

    /// Cabecera plegable: la pastilla con el modo y, desplegada, el PC, el
    /// Teclado, los chips, el selector «En Cemu soy» y «Salir». Ya no hay
    /// puertas de anchura: en la tarjeta cabe todo en cualquier móvil.
    private var header: some View {
        CollapsibleHeader(handleLabel: headerLabel, connected: link.link.connected != nil) {
            VStack(spacing: 8) {
                HStack(spacing: 20) {
                    if let c = link.link.connected {
                        Text(c.pcName).pepoBody().lineLimit(1).frame(maxWidth: 160)
                        if c.hasKeyboard && (c.mode == LinkState.modeCemu || c.mode == LinkState.modeRetroArch) {
                            HeaderKeyboardButton { keyboardOpen = true }
                        }
                    }
                    // Reconectando, el punto latiendo: es la única señal de que
                    // el servicio está rehaciendo la sesión (como en Android)
                    if case .reconnecting(let pc, _) = link.link {
                        ReconnectingLabel(pcName: pc, font: PepoFont.bodyMedium())
                    }
                    TextLink(title: tr("exit"), color: Pepo.error, action: onDisconnect).fixedSize()
                }
                if let c = link.link.connected {
                    if showModeChips(c, showChips) {
                        ModeChips(current: c.mode, supportsCemu: c.supportsCemu, supportsSwitch: c.supportsSwitch, supportsRetroArch: c.supportsRetroArch, supportsPointer: c.receiver.supportsPointer, compact: true)
                    }
                    // En Dolphin, el chip «Nunchuk» (aquí apagado: encenderlo
                    // cambia este NES por el mando + Nunchuk)
                    if showNunchukChip(c) { NunchukChip(link: c, compact: true) }
                    if isWiiUAsWiimote(c) || Route.isRetroArch(link.link) {
                        PadSelector(link: c, width: HeaderCollapse.cardWidth - 40, compact: true)
                    }
                    if Route.isRetroArch(link.link) {
                        RetroArchHotkeys(compact: true)
                    }
                }
            }
        }
    }

    /// Texto de la pastilla: lo mismo que decía la cabecera de siempre.
    private var headerLabel: String {
        switch link.link {
        case .reconnecting(let pc, _): return tr("status_reconnecting", pc)
        case .connected(let c): return isWiiUAsWiimote(c) ? tr("wiiu_as_wiimote") : modeLabel(c.mode)
        case .connecting: return tr("status_connecting")
        default: return tr("status_disconnected")
        }
    }

    private func applyEngine() {
        link.motion?.rotation = Route.sidewaysRotation(link.link, displayRotation: rotation)
    }
}

/// «Teclado» dentro de la cabecera plegable: al abrirlo, la tarjeta se pliega
/// (el teclado tapa la pantalla y al volver el mando queda despejado).
private struct HeaderKeyboardButton: View {
    @Environment(\.headerCollapse) private var header
    let onOpen: () -> Void

    var body: some View {
        KeyboardButton(compact: true) {
            header.collapse()
            onOpen()
        }
        .fixedSize()
    }
}
